rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -206,23 +232,32 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
-        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
-
-        boolean contains = receivedMessages.contains(encodedMessage);
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return true;
+        }
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
+        synchronized (receivedPeerMessages) {
+            long currentTime = System.currentTimeMillis();
+            if (currentTime - cleanMsgTimestamp > RECEIVED_MESSAGES_CACHE_DURATION) {
+                logger.trace("Cleaning {} messages from rlp queue", receivedPeerMessages.size());
+                receivedPeerMessages.clear();
+                cleanMsgTimestamp = currentTime;
             }
+        }
+
+        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
+        ReceivedPeerMessageKey receivedPeerMessageKey = new ReceivedPeerMessageKey(sender.getPeerNodeID(), encodedMessage);
 
-        } else {
+        if (receivedPeerMessages.containsKey(receivedPeerMessageKey)) {
             reportEventToPeerScoring(sender, EventType.REPEATED_MESSAGE, "Received repeated message on {}, not added to the queue");
+            return false;
         }
 
-        return !contains;
+        if (Stream.of(MessageType.BLOCK_MESSAGE, MessageType.TRANSACTIONS).anyMatch(t -> message.getMessageType() == t)) {
+            this.receivedPeerMessages.put(receivedPeerMessageKey, Instant.now().getEpochSecond());
```

Comment:
```suggestion
            this.receivedPeerMessages.put(receivedPeerMessageKey, System.currentTimeMillis());
```
I'd rather use milliseconds - they are more accurate